Core/Parser/SchemaParser.cs

```diff
@@ -675,10 +697,214 @@ private Literal ParseLiteral(TypeBase type)
+        /// <para name="serviceIndex">Index of this function within the service.</para>
+        /// <returns>The parsed rpc function definition.</returns>
+        private FunctionDefinition? ParseFunctionDefinition(string serviceName, string serviceIndex)
+        {
+            var definitionDocumentation = ConsumeBlockComments();
+            
+            // The start of the function is the definition token of the function
+            var definitionToken = CurrentToken;
+            var functionStart = CurrentToken.Span;
+
+            var isReadonly = Eat(TokenKind.ReadOnly);
+            var returnType = EatPseudoKeyword("void") ? null : ParseType(definitionToken);
+            var returnTypeSpan = functionStart.Combine(CurrentToken.Span);
+
+            const string hint = "A function must be defined with a return type, name, and arguments such as 'RetType myFunction(ArgType arg1, OtherArg arg2);' or 'void emptyFn(void);'";
+            var name = ExpectLexeme(TokenKind.Identifier, hint);
+            
+            Expect(TokenKind.OpenParenthesis, hint);
+
+            var argList = new List<Field>();
+            var argsStart = CurrentToken.Span;
+            if (EatPseudoKeyword("void"))
+            {
+                Expect(TokenKind.CloseParenthesis, hint);
+            }
+            else
+            {
+                // read parameter list
+                while (true)
+                {
+                    var paramStart = CurrentToken.Span;
+                    var paramType = ParseType(definitionToken);
+                    var paramName = ExpectLexeme(TokenKind.Identifier, hint);
+                    var paramSpan = paramStart.Combine(CurrentToken.Span);
+                    
+                    if (argList.Any(t => t.Name.Equals(paramName)))
+                    {
+                        _errors.Add(new DuplicateArgumentName(definitionToken.Span.Combine(CurrentToken.Span), serviceName, serviceIndex, paramName));
+                    }
+                    else
+                    {
+                        argList.Add(new Field(paramName, paramType, paramSpan, null, 0, ""));
+                    }
+                    
+                    if (Eat(TokenKind.CloseParenthesis))
+                    {
+                        break;
+                    }
+                    Expect(TokenKind.Comma, "Function arguments must be separated by commas");
+                }
+            }
+            var argsSpan = argsStart.Combine(CurrentToken.Span);
+            
+            Expect(TokenKind.Semicolon, "Function definition must end with a ';' semicolon");
+
+            var functionSpan = functionStart.Combine(CurrentToken.Span);
+            
+            StartScope();
+            var indexAsOpcode = new OpcodeAttribute(serviceIndex, true);
+            var returnStruct = new StructDefinition(
+                $"_{serviceName.ToPascalCase()}{name.ToPascalCase()}Return",
```

Comment:
I think we should throw an extra delimiter between the service name and function name for good measure, so that `_User_ListAdd` and `_UserList_Add` or something won't clash.

Reply:
It might be cool to make some helper methods for this, like

```cs
private static string RpcNamePrefix(string serviceName, string name)
    => $"_{serviceName.ToPascalCase()}_{name.ToPascalCase()}";

private static string RpcReturnStructName(string serviceName, string name)
    => RpcNamePrefix(serviceName, name) + "Return";
```

just in case we change our minds about e.g. the Pascal-casing.

Reply:
Two things on this
1. I assume that the different languages will convert to whatever name capitalization makes sense for them (e.g. Rust will use pascal for structs no matter what the name here is and snake case for functions) if they care
2. You are right that another delimiter is technically needed, but a `_` might be problematic because of the previous point.

Since the structs are not really to be used by the end user we could just uglify them a bit and use something like "_Service{serviceName}Function{functionName}" even. All that matters is we know what it is when we go to generate code.